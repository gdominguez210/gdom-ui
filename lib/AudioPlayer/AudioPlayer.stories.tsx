import type { StoryObj, Meta } from '@storybook/react';
import { AudioPlayer as AudioPlayerComponent } from '@lib/AudioPlayer';
import type { AudioTrackData } from '@lib/AudioPlayerContextProvider';

export default {
  title: 'components/AudioPlayer',
  component: AudioPlayerComponent,
  subcomponents: {
    AudioPlayerAuthor: AudioPlayerComponent.Author,
    AudioPlayerContextProvider: AudioPlayerComponent.ContextProvider,
    AudioPlayerControls: AudioPlayerComponent.Controls,
    AudioPlayerImage: AudioPlayerComponent.Image,
    AudioPlayerInfo: AudioPlayerComponent.Info,
    AudioPlayerProgressBar: AudioPlayerComponent.ProgressBar,
    AudioPlayerTime: AudioPlayerComponent.Time,
    AudioPlayerTitle: AudioPlayerComponent.Title,
    AudioPlayerVolume: AudioPlayerComponent.Volume,
  },
} as Meta<typeof AudioPlayerComponent>;

const data: AudioTrackData[] = [
  {
    title: 'Electro Swing House',
    src: 'https://cdn1.suno.ai/2fdedd0d-a05b-4dda-8c8d-09ba8697a50f.mp3',
    author: 'Gary Dominguez',
    thumbnail: 'https://cdn2.suno.ai/image_7fa89cca-b468-4ec4-9649-70dd745acd12.jpeg',
  },
  {
    title: 'K-Block (Demo)',
    src: 'https://cdn1.suno.ai/a9ef09b8-fbc1-4bd4-ac0c-4a9e82c561ca.mp3',
    author: 'Gary Dominguez',
    thumbnail: 'https://cdn2.suno.ai/image_a9ef09b8-fbc1-4bd4-ac0c-4a9e82c561ca.jpeg',
  },
  {
    title: 'Doomcore',
    src: 'https://cdn1.suno.ai/e5130783-0f8b-4496-8fcd-ae58c0fa02f2.mp3',
    author: 'Gary Dominguez',
    thumbnail: 'https://cdn2.suno.ai/image_ca4a5344-69d6-4967-87a5-324825d0ebaf.jpeg',
  },
  {
    title: 'Doomcore V2',
    src: 'https://cdn1.suno.ai/0ec8d8ef-3f08-4d1a-a7d4-b51536e9b1a7.mp3',
    author: 'Gary Dominguez',
    thumbnail: 'https://cdn2.suno.ai/image_6b1d56d3-c2aa-4b5d-b759-ff9fa2055746.jpeg',
  },
  {
    title: 'Adrift in Space',
    src: 'https://cdn1.suno.ai/f4e9bdb8-04c6-4e98-9ea2-8b4bce21156e.mp3',
    author: 'Gary Dominguez',
    thumbnail: 'https://cdn2.suno.ai/image_7a168d5e-ebbe-470a-a5cd-aebe3318204c.jpeg',
  },
  {
    title: 'The Awakening',
    src: 'https://cdn1.suno.ai/582c6984-bbfd-44cb-a2df-6be4e4cdadf8.mp3',
    author: 'Gary Dominguez',
    thumbnail: 'https://cdn2.suno.ai/image_0769ed7e-5990-4b23-986a-2ca3281e1ac6.jpeg',
  },
] as const;

export const AudioPlayer: StoryObj<typeof AudioPlayerComponent> = {
  args: {},
  render: () => (
    <AudioPlayerComponent.ContextProvider tracks={data}>
      <AudioPlayerComponent>
        <div className="flex flex-grow justify-space-between gap-4">
          <AudioPlayerComponent.Info className="basis-1/3">
            <AudioPlayerComponent.Image />
            <div className="py-2">
              <AudioPlayerComponent.Title />
              <AudioPlayerComponent.Author />
            </div>
          </AudioPlayerComponent.Info>
          <AudioPlayerComponent.Controls />
          <AudioPlayerComponent.Volume className="ml-auto pr-4" />
        </div>
        <AudioPlayerComponent.ProgressBar />
      </AudioPlayerComponent>
    </AudioPlayerComponent.ContextProvider>
  ),
};
