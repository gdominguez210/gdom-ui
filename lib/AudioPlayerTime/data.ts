import { intervalToDuration } from 'date-fns';

function isDefined<T>(argument: T | undefined): argument is T {
  return argument !== undefined;
}

export function formatAudioDurationForDisplay(audioDurationInSeconds: number = 0) {
  const {
    hours,
    minutes = 0,
    seconds = 0,
  } = intervalToDuration({
    start: 0,
    end: audioDurationInSeconds * 1000,
  });

  const formatWithZero = (num: number) => String(num).padStart(2, '0');

  return [hours, minutes, seconds].filter(isDefined).map(formatWithZero).join(':');
}
