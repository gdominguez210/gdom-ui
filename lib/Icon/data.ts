import StarLine from '@lib/assets/svgs/star-line.svg';
import ForwardEndFill from '@lib/assets/svgs/forward-end-fill.svg';
import PauseLargeFill from '@lib/assets/svgs/pause-large-fill.svg';
import PlayLargeFill from '@lib/assets/svgs/play-large-fill.svg';
import RewindFill from '@lib/assets/svgs/rewind-fill.svg';
import RewindStartFill from '@lib/assets/svgs/rewind-start-fill.svg';
import ShuffleFill from '@lib/assets/svgs/shuffle-fill.svg';
import SpeedFill from '@lib/assets/svgs/speed-fill.svg';
import StopLargeFill from '@lib/assets/svgs/stop-large-fill.svg';
import VolumeMuteFill from '@lib/assets/svgs/volume-mute-fill.svg';
import VolumeDownFill from '@lib/assets/svgs/volume-down-fill.svg';
import VolumeUpFill from '@lib/assets/svgs/volume-up-fill.svg';
import RepeatFill from '@lib/assets/svgs/repeat-fill.svg';
import RepeatOneFill from '@lib/assets/svgs/repeat-one-fill.svg';
import Repeat2Fill from '@lib/assets/svgs/repeat-2-fill.svg';
import DiscFill from '@lib/assets/svgs/disc-fill.svg';

export const icons = {
  'star-line': StarLine,
  'forward-end-fill': ForwardEndFill,
  'pause-large-fill': PauseLargeFill,
  'play-large-fill': PlayLargeFill,
  'repeat-fill': RepeatFill,
  'repeat-one-fill': RepeatOneFill,
  'repeat-2-fill': Repeat2Fill,
  'rewind-fill': RewindFill,
  'rewind-start-fill': RewindStartFill,
  'shuffle-fill': ShuffleFill,
  'speed-fill': SpeedFill,
  'stop-large-fill': StopLargeFill,
  'volume-mute-fill': VolumeMuteFill,
  'volume-down-fill': VolumeDownFill,
  'volume-up-fill': VolumeUpFill,
  'disc-fill': DiscFill,
} as const;

export type IconName = keyof typeof icons;
