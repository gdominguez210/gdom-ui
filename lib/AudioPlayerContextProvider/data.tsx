import type { ReactNode } from 'react';
import { trackData } from '@lib/AudioPlayer/data';
import { AudioPlayerContextProvider } from '@lib/AudioPlayerContextProvider/AudioPlayerContextProvider';

export const AUDIO_PLAYER_CONTEXT_ERROR =
  'useAudioPlayerContext must be used within an AudioPlayerContextProvider';

export const AudioPlayerContextProviderWithTrackData = ({ children }: { children: ReactNode }) => (
  <AudioPlayerContextProvider tracks={trackData}>{children}</AudioPlayerContextProvider>
);
