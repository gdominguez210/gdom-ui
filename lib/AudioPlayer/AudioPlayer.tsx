import { type HTMLAttributes, type ElementType } from 'react';
import clsx from 'clsx';
import { twMerge } from 'tailwind-merge';
import { AudioPlayerAuthor } from '@lib/AudioPlayerAuthor';
import { AudioPlayerContextProvider } from '@lib/AudioPlayerContextProvider';
import { AudioPlayerControls } from '@lib/AudioPlayerControls';
import { AudioPlayerImage } from '@lib/AudioPlayerImage';
import { AudioPlayerInfo } from '@lib/AudioPlayerInfo';
import { AudioPlayerProgressBar } from '@lib/AudioPlayerProgressBar';
import { AudioPlayerTime } from '@lib/AudioPlayerTime';
import { AudioPlayerTitle } from '@lib/AudioPlayerTitle';
import { AudioPlayerVolume } from '@lib/AudioPlayerVolume';

export interface AudioPlayerProps extends HTMLAttributes<HTMLElement> {
  /** @default div */
  as?: ElementType;
}

export function AudioPlayer(props: AudioPlayerProps) {
  const { as = 'div', children, className } = props;

  const Node = as;

  return (
    <Node
      className={twMerge(
        clsx('flex flex-col justify-center bg-slate-700 text-neutral-100', className),
      )}
    >
      {children}
    </Node>
  );
}

AudioPlayer.Author = AudioPlayerAuthor;
AudioPlayer.ContextProvider = AudioPlayerContextProvider;
AudioPlayer.Controls = AudioPlayerControls;
AudioPlayer.Image = AudioPlayerImage;
AudioPlayer.Info = AudioPlayerInfo;
AudioPlayer.ProgressBar = AudioPlayerProgressBar;
AudioPlayer.Time = AudioPlayerTime;
AudioPlayer.Title = AudioPlayerTitle;
AudioPlayer.Volume = AudioPlayerVolume;
