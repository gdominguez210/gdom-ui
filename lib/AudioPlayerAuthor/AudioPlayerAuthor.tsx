import { type HTMLAttributes, type ElementType } from 'react';
import clsx from 'clsx';
import { twMerge } from 'tailwind-merge';
import { useAudioPlayerContext } from '@lib/AudioPlayerContextProvider/useAudioPlayerContext';

export interface AudioPlayerAuthorProps extends HTMLAttributes<HTMLElement> {
  /** @default p */
  as?: ElementType;
}

export function AudioPlayerAuthorBase(props: AudioPlayerAuthorProps) {
  const { as = 'p', className, children, ...restProps } = props;

  const Node = as;

  return (
    <Node
      className={twMerge(clsx('line-clamp-1 text-sm text-gray-400', className))}
      {...restProps}
    >
      {children}
    </Node>
  );
}

export function AudioPlayerAuthor(props: AudioPlayerAuthorProps) {
  const { currentTrack: { author } = {} } = useAudioPlayerContext();

  if (!author) return null;

  return (
    <AudioPlayerAuthorBase
      {...props}
      title={author}
    >
      {author}
    </AudioPlayerAuthorBase>
  );
}
