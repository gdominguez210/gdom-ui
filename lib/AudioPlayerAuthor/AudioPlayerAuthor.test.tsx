import { screen, render, waitFor } from '@testing-library/react';
import { describe, expect, test, vi } from 'vitest';
import { AudioPlayerAuthor } from '@lib/AudioPlayerAuthor/AudioPlayerAuthor';
import { trackData } from '@lib/AudioPlayer/data';
import {
  AudioPlayerContextProviderWithTrackData as AudioWrapper,
  AUDIO_PLAYER_CONTEXT_ERROR,
} from '@lib/AudioPlayerContextProvider/data';

describe('AudioPlayerAuthor should...', () => {
  test('match the snapshot', () => {
    const { container } = render(<AudioPlayerAuthor />, { wrapper: AudioWrapper });

    expect(container).toMatchSnapshot();
  });

  test('should throw an error when not used in AudioPlayerContextProvider', async () => {
    vi.spyOn(console, 'error').mockImplementation(() => vi.fn());

    await waitFor(() =>
      expect(() => render(<AudioPlayerAuthor />)).toThrow(AUDIO_PLAYER_CONTEXT_ERROR),
    );
    vi.restoreAllMocks();
  });

  test("display a track's author", () => {
    render(<AudioPlayerAuthor data-testid="audio-player-author" />, { wrapper: AudioWrapper });

    const element = screen.getByTestId('audio-player-author');

    expect(element).toHaveTextContent(trackData[0]?.author as string);
  });

  test('allow the user to change the returned element type', () => {
    render(
      <AudioPlayerAuthor
        as="span"
        data-testid="audio-player-author"
      />,
      { wrapper: AudioWrapper },
    );

    const element = screen.getByTestId('audio-player-author');

    expect(element.nodeName.toLowerCase()).toBe('span');
  });

  test('allow the user to add a className', () => {
    render(
      <AudioPlayerAuthor
        className="text-md"
        data-testid="audio-player-author"
      />,
      { wrapper: AudioWrapper },
    );

    const element = screen.getByTestId('audio-player-author');

    expect(element).toHaveClass('text-md');
  });
});
