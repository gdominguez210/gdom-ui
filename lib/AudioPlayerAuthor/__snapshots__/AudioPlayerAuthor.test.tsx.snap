// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AudioPlayerAuthor should... > match the snapshot 1`] = `
<div>
  <p
    class="line-clamp-1 text-sm text-gray-400"
    title="Gary Dominguez"
  >
    Gary Dominguez
  </p>
</div>
`;
